File: extensions/android/audio-processing/src/external_audio_processing_factory.cc
/*
 * Copyright 2022 LiveKit
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "extensions/android/audio-processing/src/external_audio_processing_factory.h"

#include <jni.h>
#include <syslog.h>

#include "api/audio/builtin_audio_processing_builder.h"
#include "api/environment/environment_factory.h"
#include "api/make_ref_counted.h"
#include "rtc_base/ref_counted_object.h"
#include "extensions/android/audio-processing/generated_effects_jni/ExternalAudioProcessingFactory_jni.h"
#include "sdk/android/native_api/jni/java_types.h"
#include "sdk/android/native_api/jni/scoped_java_ref.h"
#include "sdk/android/src/jni/jni_helpers.h"
#include "extensions/android/audio-processing/src/external_audio_processor.h"

namespace webrtc {
namespace jni {

ExternalAudioProcessingJni::ExternalAudioProcessingJni(
    JNIEnv* jni,
    const JavaRef<jobject>& j_processing)
    : j_processing_global_(jni, j_processing) {}
ExternalAudioProcessingJni::~ExternalAudioProcessingJni() {}
void ExternalAudioProcessingJni::Initialize(int sample_rate_hz,
                                            int num_channels) {
  JNIEnv* env = AttachCurrentThreadIfNeeded();
  Java_AudioProcessing_initialize(env, j_processing_global_, sample_rate_hz,
                                  num_channels);
}

void ExternalAudioProcessingJni::Reset(int new_rate) {
  JNIEnv* env = AttachCurrentThreadIfNeeded();
  Java_AudioProcessing_reset(env, j_processing_global_, new_rate);
}

void ExternalAudioProcessingJni::Process(int num_bands, int num_frames, int buffer_size, float* buffer) {
  JNIEnv* env = AttachCurrentThreadIfNeeded();
  ScopedJavaLocalRef<jobject> audio_buffer =
      NewDirectByteBuffer(env, (void*)buffer, buffer_size * sizeof(float));
  Java_AudioProcessing_process(env, j_processing_global_, num_bands, num_frames, audio_buffer);
}

ExternalAudioProcessingFactory::ExternalAudioProcessingFactory() {
  capture_post_processor_ = new ExternalAudioProcessor();
  std::unique_ptr<webrtc::CustomProcessing> capture_post_processor(
      capture_post_processor_);

  render_pre_processor_ = new ExternalAudioProcessor();
  std::unique_ptr<webrtc::CustomProcessing> render_pre_processor(
      render_pre_processor_);

  apm_ = webrtc::BuiltinAudioProcessingBuilder()
             .SetCapturePostProcessing(std::move(capture_post_processor))
             .SetRenderPreProcessing(std::move(render_pre_processor))
             .Build(CreateEnvironment());

  webrtc::AudioProcessing::Config config;
  apm_->ApplyConfig(config);
}

static webrtc::scoped_refptr<ExternalAudioProcessingFactory> default_processor;

static jlong JNI_ExternalAudioProcessingFactory_GetDefaultApm(JNIEnv* env) {
  if (!default_processor) {
    default_processor = webrtc::make_ref_counted<ExternalAudioProcessingFactory>();
  }
  return webrtc::jni::jlongFromPointer(default_processor->apm().get());
}

static void JNI_ExternalAudioProcessingFactory_SetCaptureCompressionGain(
    JNIEnv* env,
    jint compression_gain_db) {
  if (!default_processor) {
    return;
  }
  default_processor->apm().get()->SetRuntimeSetting(AudioProcessing::RuntimeSetting::CreateCompressionGainDb(compression_gain_db));
}


static void JNI_ExternalAudioProcessingFactory_EnableGainController1(
    JNIEnv* env,
    jint mode,
    jint target_level_dbfs,
    jint compression_gain_db,
    jboolean enable_limiter
  ) {
  if (!default_processor) {
    return;
  }
  if (!default_processor->apm()) {
    return;
  }
  webrtc::AudioProcessing::Config config = default_processor->apm()->GetConfig();
  config.gain_controller1.enabled = true;
  config.gain_controller2.enabled = false;
  if (mode == 0) {
    config.gain_controller1.mode = webrtc::AudioProcessing::Config::GainController1::kAdaptiveAnalog;
  } else if (mode == 1) {
    config.gain_controller1.mode = webrtc::AudioProcessing::Config::GainController1::kAdaptiveDigital;
  } else if (mode == 2) {
    config.gain_controller1.mode = webrtc::AudioProcessing::Config::GainController1::kFixedDigital;
  }
  config.gain_controller1.target_level_dbfs = target_level_dbfs;
  config.gain_controller1.compression_gain_db = compression_gain_db;
  config.gain_controller1.enable_limiter = enable_limiter;
  default_processor->apm()->ApplyConfig(config);
}

static jlong JNI_ExternalAudioProcessingFactory_SetCapturePostProcessing(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_processing) {
  if (!default_processor) {
    return 0;
  }
  auto processing =
      webrtc::make_ref_counted<ExternalAudioProcessingJni>(env, j_processing);
  processing->AddRef();
  default_processor->capture_post_processor()->SetExternalAudioProcessing(
      processing.get());
  return jlongFromPointer(processing.get());
}

static jlong JNI_ExternalAudioProcessingFactory_SetRenderPreProcessing(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_processing) {
  if (!default_processor) {
    return 0;
  }
  auto processing =
      webrtc::make_ref_counted<ExternalAudioProcessingJni>(env, j_processing);
  processing->AddRef();
  default_processor->render_pre_processor()->SetExternalAudioProcessing(
      processing.get());
  return jlongFromPointer(processing.get());
}

static void JNI_ExternalAudioProcessingFactory_SetBypassFlagForCapturePost(
    JNIEnv* env,
    jboolean bypass) {
  if (!default_processor) {
    return;
  }
  default_processor->capture_post_processor()->SetBypassFlag(bypass);
}

static void JNI_ExternalAudioProcessingFactory_SetBypassFlagForRenderPre(
    JNIEnv* env,
    jboolean bypass) {
  if (!default_processor) {
    return;
  }
  default_processor->render_pre_processor()->SetBypassFlag(bypass);
}

static void JNI_ExternalAudioProcessingFactory_Destroy(JNIEnv* env) {
  if (!default_processor) {
    return;
  }
  default_processor->render_pre_processor()->SetExternalAudioProcessing(
      nullptr);
  default_processor->capture_post_processor()->SetExternalAudioProcessing(
      nullptr);
  default_processor = nullptr;
}

}  // namespace jni
}  // namespace webrtc

File: extensions/android/audio-processing/java/org/webrtc/ExternalAudioProcessingFactory.java
/*
 * Copyright 2022 LiveKit
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.webrtc;

import java.nio.ByteBuffer;

import androidx.annotation.Nullable;
import org.webrtc.AudioProcessingFactory;


public class ExternalAudioProcessingFactory implements AudioProcessingFactory {

  /**
   * Interface for external audio processing.
   */
  public static interface AudioProcessing {
    /**
     * Called when the processor should be initialized with a new sample rate and
     * number of channels.
     */
    @CalledByNative("AudioProcessing")
    void initialize(int sampleRateHz, int numChannels);
    /** Called when the processor should be reset with a new sample rate. */  
    @CalledByNative("AudioProcessing")
    void reset(int newRate);
    /**  
     * Processes the given capture or render signal. NOTE: `buffer.data` will be  
     * freed once this function returns so callers who want to use the data  
     * asynchronously must make sure to copy it first.  
     */
    @CalledByNative("AudioProcessing")
    void process(int numBands, int numFrames, ByteBuffer buffer);
  }

  private long apmPtr;
  private long capturePostProcessingPtr;
  private long renderPreProcessingPtr;

  public ExternalAudioProcessingFactory() {
    apmPtr = nativeGetDefaultApm();
    capturePostProcessingPtr = 0;
    renderPreProcessingPtr = 0;
  }

  @Override
  public long createNative(long webrtcEnvRef) {
    if(apmPtr == 0) {
      apmPtr = nativeGetDefaultApm();
    }
    return apmPtr;
  }

  /**
   * Sets the capture post processing module. 
   * This module is applied to the audio signal after capture and before sending 
   * to the audio encoder.
   */
  public void setCapturePostProcessing(@Nullable AudioProcessing processing) {
    checkExternalAudioProcessorExists();
    long newPtr = nativeSetCapturePostProcessing(processing);
    if (capturePostProcessingPtr != 0) {
      JniCommon.nativeReleaseRef(capturePostProcessingPtr);
      capturePostProcessingPtr = 0;
    }
    capturePostProcessingPtr = newPtr;
  }

  /**
   * Runtime sets the compressionGainDb for the capture audio processing module.
   * ranging [0, 90].
   */
  public void setCaptureCompressionGain(int compressionGainDb) {
    checkExternalAudioProcessorExists();
    nativeSetCaptureCompressionGain(compressionGainDb);
  }

  /**
   * Sets the compressionGainDb for the capture audio processing module.
   * targetLevelDbfs: default(3)
   * compressionGainDb: ranging [0, 90].
   */
  public void enableGainController1(Agc1Mode mode, int targetLevelDbfs, int compressionGainDb, boolean enableLimiter) {
    checkExternalAudioProcessorExists();
    nativeEnableGainController1(mode.ordinal(), targetLevelDbfs, compressionGainDb, enableLimiter);
  }

  /**
   * Sets the render pre processing module.
   * This module is applied to the audio signal after receiving from the audio
   * decoder and before rendering.
   */
  public void setRenderPreProcessing(@Nullable AudioProcessing processing) {
    checkExternalAudioProcessorExists();
    long newPtr = nativeSetRenderPreProcessing(processing);
    if (renderPreProcessingPtr != 0) {
      JniCommon.nativeReleaseRef(renderPreProcessingPtr);
      renderPreProcessingPtr = 0;
    }
    renderPreProcessingPtr = newPtr;
  }
  
  /**
   * Sets the bypass flag for the capture post processing module.
   * If true, the registered audio processing will be bypassed.
   */
  public void setBypassFlagForCapturePost( boolean bypass) {
    checkExternalAudioProcessorExists();
    nativeSetBypassFlagForCapturePost(bypass);
  }

  /**
   * Sets the bypass flag for the render pre processing module.
   * If true, the registered audio processing will be bypassed.
   */
  public void setBypassFlagForRenderPre( boolean bypass) {
    checkExternalAudioProcessorExists();
    nativeSetBypassFlagForRenderPre(bypass);
  }

  /**
   * Destroys the ExternalAudioProcessor.
   */
  public void destroy() {
    checkExternalAudioProcessorExists();
    if (renderPreProcessingPtr != 0) {
      JniCommon.nativeReleaseRef(renderPreProcessingPtr);
      renderPreProcessingPtr = 0;
    }
    if (capturePostProcessingPtr != 0) {
      JniCommon.nativeReleaseRef(capturePostProcessingPtr);
      capturePostProcessingPtr = 0;
    }
    nativeDestroy();
    apmPtr = 0;
  }

  private void checkExternalAudioProcessorExists() {
    if (apmPtr == 0) {
      throw new IllegalStateException("ExternalAudioProcessor has been disposed.");
    }
  }

  public enum Agc1Mode {
    // Adaptive mode intended for use if an analog volume control is
    // available on the capture device. It will require the user to provide
    // coupling between the OS mixer controls and AGC through the
    // stream_analog_level() functions.
    // It consists of an analog gain prescription for the audio device and a
    // digital compression stage.
    ADAPTIVE_ANALOG,
    // Adaptive mode intended for situations in which an analog volume
    // control is unavailable. It operates in a similar fashion to the
    // adaptive analog mode, but with scaling instead applied in the digital
    // domain. As with the analog mode, it additionally uses a digital
    // compression stage.
    ADAPTIVE_DIGITAL,
    // Fixed mode which enables only the digital compression stage also used
    // by the two adaptive modes.
    // It is distinguished from the adaptive modes by considering only a
    // short time-window of the input signal. It applies a fixed gain
    // through most of the input level range, and compresses (gradually
    // reduces gain with increasing level) the input signal at higher
    // levels. This mode is preferred on embedded devices where the capture
    // signal level is predictable, so that a known gain can be applied.
    FIXED_DIGITAL
  }

  private static native void nativeSetCaptureCompressionGain(int compressionGainDb);
  private static native void nativeEnableGainController1(int mode, int targetLevelDbfs, int compressionGainDb, boolean enableLimiter);
  private static native long nativeGetDefaultApm();
  private static native long nativeSetCapturePostProcessing(AudioProcessing processing);
  private static native long nativeSetRenderPreProcessing(AudioProcessing processing);
  private static native void nativeSetBypassFlagForCapturePost(boolean bypass);
  private static native void nativeSetBypassFlagForRenderPre(boolean bypass);
  private static native void nativeDestroy();
}
